### DISCUSSION TITLE
Helix

### DISCUSSION THREAD

**Discussion by jgomez720** [KittyCAD/modeling-app#1048] [New Feature Requests]
### Description
A tool that allows the user to create a helix on a cylindrical surface.

### Functionality
1. User selects the _Helix_ button.
2. User selects a cylindrical surface
3. User inputs the amount of turns or pitch they would like
4. User selects clockwise or counter-clockwise
5. User selects the starting location (can be set by an angle, similar to what's seen in the video below)
6. **Optional:** user selects a length (only if they don't want the helix to extend the full length of the cylinder.

### Supporting Docs/Images
https://github.com/KittyCAD/modeling-app/assets/114548659/f0633820-1cff-42e5-a21d-2b580b5e8c62
https://github.com/KittyCAD/modeling-app/assets/114548659/8d222bef-f229-434b-b04b-0c816e7380b1

## PseudoKCL
```
const cylinderBody = startSketchOn('XY')
  |> startProfileAt([4, 0], %)
  |> arc({
     angle_end: 360,
     angle_start: 0,
     radius: 4
   }, %)
  |> close(%)
  |> extrude(15, %)

const helixPath = helix({
  surface: cylinderBody,
  revolutions: 16,
  angle_start: 0,
  // ccw: true/false. Optional argument if you want it to be counter clockwise. Default is clockwise.
  // length: ____ . Optional argument if the user does not want it to extend the full length of the cylinder surface.
})
```

**Expected Result:**
![Screenshot 2023-11-09 at 4 18 16 PM](https://github.com/KittyCAD/modeling-app/assets/114548659/3c83c37e-8f9c-4b66-8468-dda594b72efb)

**Comment by Irev-Dev** [comment]
So I think we need to think through a few more use cases for this.
For one passing in a cylinder is a nice thing to have because we can drive a bunch from it, like the length of helix and it's diameter, but I still think we need a raw helix, if you want to model a spring you will have to add a cylinder just to remove it again later.

But even if the cylinder, I think we need one that takes pitch, I know it's not hard to calculate pitch if you know the revolutions and length, but I think pitch is probably going to be used more, at least for things like thread, pitch is much more important that total revolutions.

So maybe we have 3 functions

```ts
helixRaw({ // need a better name.
  length: 20,
  pitch: 1,
  radius: 2.5,
  plane: planeData, // could use startSketchOn name is no great but does return a plane
  // angleStart: 0,
  // ccw: true,
})

helixPitch({
  surface: cylinderBody,
  pitch: 2.5,
  // angleStart: 0,
  // ccw: true,
})

helixRevolutions({
  surface: cylinderBody,
  revolutions: 10,
  // angleStart: 0,
  // ccw: true,
})
```

Or alternatively we keep just the one helix function, but helper function for that can take the cylinder id and return the plane data or similar that's needed.

```ts
helix({
  length: 20,
  pitch: 1,
  radius: 2.5,
  plane: planeData, // could use startSketchOn name is no great but does return a plane
  // angleStart: 0,
  // ccw: true,
})

helix(helixPitchHelper({
  surface: cylinderBody,
  pitch: 2.5,
  // angleStart: 0,
  // ccw: true,
}))

helix(helixRevolutionsHelper({
  surface: cylinderBody,
  revolutions: 10,
  // angleStart: 0,
  // ccw: true,
}))
```

Thoughts?

**Reply by lf94** [reply]
In my code CAD experience the latter is better because it keeps the API surface small, it keeps the behavior easy to reason about (you can inspect the values the helper is generating!), you can create your own helper (!!), and API updates are way easier (older helper can stay and a new helper is created).

**Reply by jgomez720** [reply]
I agree with @lf94, I like the second more

**Reply by jessfraz** [reply]
but like why do we need helixPitchHelper when we could just have helix accept `{
  length: 20,
  pitch: 1,
  radius: 2.5,
  plane: planeData, // could use startSketchOn name is no great but does return a plane
  // angleStart: 0,
  // ccw: true,
}` or  `{
  surface: cylinderBody,
  revolutions: 10,
  // angleStart: 0,
  // ccw: true,
}`

**Reply by jessfraz** [reply]
like we dont need the helper we can make the object helix accepts be one of many possible objects.

**Reply by Irev-Dev** [reply]
I just realised I have `helixPitchHelper` in there twice when it should be `helixPitchHelper` and `helixRevolutionsHelper`

But I think your point still stands Jess.

two disadvantages to overloading the inputs
- Having different enums as inputs into functions is something that's only available to stdlib functions, we already do this a bunch, but maybe it's not the best to have functions with features that would never be possible to do as functions defined in the language.
- Having overloading inputs makes the ast mods slightly more onerous, instead of just checking the name of a call expression, you need to start checking the names of the input args to verify which type of helix it is.

These are relatively minor.

**Reply by jessfraz** [reply]
That’s fair another con is autocomplete only does one

**Reply by jessfraz** [reply]
So yeah I’m fine w multiple functions too
